design_docs/analysis_task_template.md

```diff
@@ -0,0 +1,594 @@
+2. ensure that tasks implement a standard set of
+   functions, making it easier to perform actions (such as checking whether
+   the task should be run, checking for required model and observations files,
+   and running the analysis) on each analysis task in sequence (and, in the
+   future, in parallel)
+3. demonstrate the syntax and style of docstrings required to comment/document
+   each task and each function
+
+<h3> Requirements </h3>
+
+<h4> Requirement: Template for Analysis Tasks <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+The template should include each function that each analysis task *must* implement
+and and example docstring both for the task as a whole and for each funciton.
+
+<h4> Requirement: Validation within Analysis Tasks <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+Validation, such as checking config options or adding new ones if they are missing,
+or checking inf required data files are present, should be performed within a
+function in each task (rather than in `run_analysis.py`, as is sometimes the current
+case).  
+
+<h4> Requirement: Analysis Continues even when Analysis Task Fails <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+If validation fails, an error message should be printed but other analysis
+tasks should be allowed to run.  Similarly, if a given analysis task raises
+an exception, the error and stack trace should be printed but other analysis
+tasks should still be run.
+
+<h4> Requirement: List of Tasks to Perform<br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+There should be a single place where new tasks are added to `run_analysis.py`, as
+is presently the case.  Yet, there should be a way to create a list of tasks to be
+performed and later determine whether, when and how those tasks are to be run.
+This change is requried to allow for later task parallelism. Currently, a task 
+module is imported, there is a check to see if that task should be run, and the
+task is performed in immediate sequence.
+
+<h3> Algorithmic Formulations (optional) </h3>
+
+<h4> Design solution: Template for Analysis Tasks <br>
+Date last modified: 2017/03/08 <br>
+Contributors: Xylar Asay-Davis
+</h4>
+The template will include the following functions:
+  - `get_task_name`: returns the name of the task
+  - `get_component_name`: returns the name of the MPAS component analyzed by the
+     task
+  - `get_category_name`: returns the category name (if any) that the task belongs to
+  - `check`: determines if the configuration is valid for running this task (e.g. if
+     necessary files and config options are present)
+  - `run`: runs the analysis task
+  
+The template will be located at:
+```
+mpas_analysis/
+    - analysis_task_template.py
+```
+That is, it is the only file (other than `__init__.py`) in the base of the 
+`mpas_analysis` directory, making it easy to find.
```

Comment:
Why put it in the root directory?  Is this to provide a "trail head" for people wishing to add a new analysis?